lightning-invoice/src/payment.rs

```diff
@@ -0,0 +1,818 @@
+	fn node_id(&self) -> PublicKey;
+
+	/// Returns the payer's channels.
+	fn first_hops(&self) -> Vec<ChannelDetails>;
+
+	/// Sends a payment over the Lightning Network using the given [`Route`].
+	fn send_payment(
+		&self, route: &Route, payment_hash: PaymentHash, payment_secret: &Option<PaymentSecret>
+	) -> Result<PaymentId, PaymentSendFailure>;
+
+	/// Retries a failed payment path for the [`PaymentId`] using the given [`Route`].
+	fn retry_payment(&self, route: &Route, payment_id: PaymentId) -> Result<(), PaymentSendFailure>;
+}
+
+/// A trait defining behavior for routing an [`Invoice`] payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` using the provided first and last hops for a
+	/// payment with the given values.
+	fn find_route(
+		&self,
+		payer: &PublicKey,
+		payee: &PublicKey,
+		payee_features: Option<InvoiceFeatures>,
+		first_hops: Option<&[&ChannelDetails]>,
+		last_hops: &[&RouteHint],
+		final_value_msat: u64,
+		final_cltv: u32,
+	) -> Result<Route, LightningError>;
+}
+
+/// An error that may occur when making a payment.
+#[derive(Debug)]
+pub enum PaymentError {
+	/// An error resulting from the provided [`Invoice`] or payment hash.
+	Invoice(&'static str),
+	/// An error occurring when finding a route.
+	Routing(LightningError),
+	/// An error occurring when sending a payment.
+	Sending(PaymentSendFailure),
+}
+
+impl<P: Deref, R, L: Deref, E> InvoicePayer<P, R, L, E>
+where
+	P::Target: Payer,
+	R: Router,
+	L::Target: Logger,
+	E: EventHandler,
+{
+	/// Creates an invoice payer that does not retry.
+	///
+	/// To enable retries, chain `new` with a call to [`with_retry_attempts`].
+	///
+	/// [`with_retry_attempts`]: Self::with_retry_attempts
+	pub fn new(payer: P, router: R, logger: L, event_handler: E) -> Self {
+		Self {
+			payer,
+			router,
+			logger,
+			event_handler,
+			invoice_cache: Mutex::new(HashMap::new()),
+			retry_attempts: 0,
+		}
+	}
+
+	/// Sets the maximum number of payment retry attempts before forwarding the
+	/// [`Event::PaymentPathFailed`] event to be handled by the decorated [`EventHandler`].
+	pub fn with_retry_attempts(self, retry_attempts: usize) -> Self {
+		Self { retry_attempts, ..self }
+	}
+
+	/// Pays the given [`Invoice`], caching it for later use in case a retry is needed.
+	pub fn pay_invoice(&self, invoice: &Invoice) -> Result<PaymentId, PaymentError> {
+		let payment_hash = PaymentHash(invoice.payment_hash().clone().into_inner());
+		let mut invoice_cache = self.invoice_cache.lock().unwrap();
+		match invoice_cache.entry(payment_hash) {
+			hash_map::Entry::Vacant(entry) => {
+				let payer = self.payer.node_id();
+				let payee = invoice.recover_payee_pub_key();
+				let payee_features = invoice.features().cloned();
+				let first_hops = self.payer.first_hops();
+				let last_hops = invoice.route_hints();
+				let final_value_msat = invoice.amount_pico_btc()
+					.ok_or(PaymentError::Invoice("amount missing"))? / 10;
```

Comment:
We should check for 0-value invoices here and also have a public function to allow paying a 0-value invoice by setting the value manually (via UI interaction with the user).